File: examples/mobile/ios/AionMCPViewModel.swift
import Foundation
import SwiftUI

/// ViewModel for AionMCP integration
@MainActor
class AionMCPViewModel: ObservableObject {
    private let repository: AionMCPRepository
    
    // MARK: - Published Properties
    
    @Published var health: HealthResponse?
    @Published var tools: [Tool] = []
    @Published var toolResult: ToolInvokeResponse?
    @Published var learningStats: LearningStatsResponse?
    @Published var insights: [Insight] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    // MARK: - Initialization
    
    init(repository: AionMCPRepository) {
        self.repository = repository
    }
    
    // MARK: - Health
    
    func checkHealth() async {
        isLoading = true
        errorMessage = nil
        
        do {
            health = try await repository.checkHealth()
        } catch {
            handleError(error)
        }
        
        isLoading = false
    }
    
    // MARK: - Tools
    
    func listTools(forceRefresh: Bool = false) async {
        isLoading = true
        errorMessage = nil
        
        do {
            tools = try await repository.listTools(forceRefresh: forceRefresh)
        } catch {
            handleError(error)
        }
        
        isLoading = false
    }
    
    func invokeTool(name: String, args: [String: Any]) async {
        isLoading = true
        errorMessage = nil
        toolResult = nil
        
        do {
            toolResult = try await repository.invokeTool(name: name, args: args)
        } catch {
            handleError(error)
        }
        
        isLoading = false
    }
    
    // MARK: - Learning
    
    func getLearningStats() async {
        isLoading = true
        errorMessage = nil
        
        do {
            learningStats = try await repository.getLearningStats()
        } catch {
            handleError(error)
        }
        
        isLoading = false
    }
    
    func getInsights(type: String? = nil, priority: String? = nil) async {
        isLoading = true
        errorMessage = nil
        
        do {
            insights = try await repository.getInsights(type: type, priority: priority)
        } catch {
            handleError(error)
        }
        
        isLoading = false
    }
    
    func getToolInsights(toolName: String) async {
        isLoading = true
        errorMessage = nil
        
        do {
            insights = try await repository.getToolInsights(toolName: toolName)
        } catch {
            handleError(error)
        }
        
        isLoading = false
    }
    
    func analyzePatterns() async {
        isLoading = true
        errorMessage = nil
        
        do {
            _ = try await repository.analyzePatterns()
        } catch {
            handleError(error)
        }
        
        isLoading = false
    }
    
    // MARK: - Cache Management
    
    func clearCache() async {
        await repository.clearCache()
        tools = []
        health = nil
        learningStats = nil
        insights = []
        toolResult = nil
    }
    
    // MARK: - Convenience Methods
    
    func refresh() async {
        await loadInitialData()
    }
    
    func loadInitialData() async {
        await checkHealth()
        await listTools()
    }
    
    // MARK: - Error Handling
    
    private func handleError(_ error: Error) {
        if let aionError = error as? AionMCPError {
            errorMessage = aionError.errorDescription
        } else {
            errorMessage = error.localizedDescription
        }
    }
}

// MARK: - Convenience Initializer

extension AionMCPViewModel {
    convenience init(baseURL: String, apiKey: String? = nil) {
        let client = AionMCPClient(baseURL: baseURL, apiKey: apiKey)
        let repository = AionMCPRepository(client: client)
        self.init(repository: repository)
    }
}

File: examples/mobile/android/AionMCPViewModel.kt
package com.example.aionmcp.ui.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.aionmcp.data.models.*
import com.example.aionmcp.data.repository.AionMCPRepository
import kotlinx.coroutines.launch

/**
 * ViewModel for AionMCP integration
 * Manages UI state and coordinates repository operations
 */
class AionMCPViewModel(
    private val repository: AionMCPRepository
) : ViewModel() {
    
    // Health status
    private val _healthStatus = MutableLiveData<Result<HealthResponse>>()
    val healthStatus: LiveData<Result<HealthResponse>> = _healthStatus
    
    // Tools list
    private val _tools = MutableLiveData<Result<List<Tool>>>()
    val tools: LiveData<Result<List<Tool>>> = _tools
    
    // Tool invocation result
    private val _toolResult = MutableLiveData<Result<ToolInvokeResponse>>()
    val toolResult: LiveData<Result<ToolInvokeResponse>> = _toolResult
    
    // Learning statistics
    private val _learningStats = MutableLiveData<Result<LearningStatsResponse>>()
    val learningStats: LiveData<Result<LearningStatsResponse>> = _learningStats
    
    // Insights
    private val _insights = MutableLiveData<Result<List<Insight>>>()
    val insights: LiveData<Result<List<Insight>>> = _insights
    
    // Loading state
    private val _isLoading = MutableLiveData<Boolean>()
    val isLoading: LiveData<Boolean> = _isLoading
    
    /**
     * Check server health
     */
    fun checkHealth() {
        viewModelScope.launch {
            _isLoading.value = true
            val result = repository.checkHealth()
            _healthStatus.value = result
            _isLoading.value = false
        }
    }
    
    /**
     * List all available tools
     */
    fun listTools(forceRefresh: Boolean = false) {
        viewModelScope.launch {
            _isLoading.value = true
            val result = repository.listTools(forceRefresh)
            _tools.value = result
            _isLoading.value = false
        }
    }
    
    /**
     * Invoke a tool
     */
    fun invokeTool(toolName: String, args: Map<String, Any>) {
        viewModelScope.launch {
            _isLoading.value = true
            val result = repository.invokeTool(toolName, args)
            _toolResult.value = result
            _isLoading.value = false
        }
    }
    
    /**
     * Get learning statistics
     */
    fun getLearningStats() {
        viewModelScope.launch {
            _isLoading.value = true
            val result = repository.getLearningStats()
            _learningStats.value = result
            _isLoading.value = false
        }
    }
    
    /**
     * Get insights
     */
    fun getInsights(type: String? = null, priority: String? = null) {
        viewModelScope.launch {
            _isLoading.value = true
            val result = repository.getInsights(type, priority)
            _insights.value = result
            _isLoading.value = false
        }
    }
    
    /**
     * Get insights for a specific tool
     */
    fun getToolInsights(toolName: String) {
        viewModelScope.launch {
            _isLoading.value = true
            val result = repository.getToolInsights(toolName)
            _insights.value = result
            _isLoading.value = false
        }
    }
    
    /**
     * Trigger pattern analysis
     */
    fun analyzePatterns() {
        viewModelScope.launch {
            _isLoading.value = true
            repository.analyzePatterns()
            _isLoading.value = false
        }
    }
    
    /**
     * Clear cache
     */
    fun clearCache() {
        repository.clearCache()
    }
}
